web_monitoring/utils.py

```diff
@@ -122,3 +88,197 @@ def get_color_palette():
+    be used in have exited.
+
+    Override the `__exit_all__(self, type, value, traceback)` method to get a
+    version of `__exit__` that is only called when exiting the top context.
+
+    As a convenience, the built-in `__enter__` returns `self`, which is fairly
+    common, so in many cases you don't need to author your own `__enter__` or
+    `__exit__` methods.
+    """
+    _context_depth = 0
+
+    def __enter__(self):
+        self._context_depth += 1
+        return self
+
+    def __exit__(self, type, value, traceback):
+        if self._context_depth > 0:
+            self._context_depth -= 1
+        if self._context_depth == 0:
+            return self.__exit_all__(type, value, traceback)
+
+    def __exit_all__(self, type, value, traceback):
+        """
+        A version of the normal `__exit__` context manager method that only
+        gets called when the top level context is exited. This is meant to be
+        overridden in your class.
+        """
+        pass
+
+
+class SessionClosedError(Exception):
+    ...
+
+
+class DisableAfterCloseSession(requests.Session):
+    """
+    A custom session object raises a :class:`SessionClosedError` if you try to
+    use it after closing it, to help identify and avoid potentially dangerous
+    code patterns. (Standard session objects continue to be usable after
+    closing, even if they may not work exactly as expected.)
+    """
+    _closed = False
+
+    def close(self, disable=True):
+        super().close()
+        if disable:
+            self._closed = True
+
+    def send(self, *args, **kwargs):
+        if self._closed:
+            raise SessionClosedError('This session has already been closed '
+                                     'and cannot send new HTTP requests.')
+
+        return super().send(*args, **kwargs)
+
+
+class Signal:
+    """
+    A context manager to handle signals from the system safely. It keeps track
+    of previous signal handlers and ensures that they are put back into place
+    when the context exits.
+
+    Parameters
+    ----------
+    signals : int or tuple of int
+        The signal or list of signals to handle.
+    handler : callable
+        A signal handler function of the same type used with `signal.signal()`.
+        See: https://docs.python.org/3.6/library/signal.html#signal.signal
+    """
+    def __init__(self, signals, handler):
+        self.handler = handler
+        self.old_handlers = {}
+        try:
+            self.signals = tuple(signals)
+        except TypeError:
+            self.signals = (signals,)
+
+    def __enter__(self):
+        for signal_type in self.signals:
+            self.old_handlers[signal_type] = signal.getsignal(signal_type)
+            signal.signal(signal_type, self.handler)
+
+        return self
+
+    def __exit__(self, type, value, traceback):
+        for signal_type in self.signals:
+            self.old_handlers[signal_type] = signal.getsignal(signal_type)
```

Comment:
Is the intent to restore the signal handlers to the state they were before? If so, I think this line should be removed. The effect here seems to be stashing the _current_ state of the handlers and then setting them to be exactly the same as they are---a no-op.

Reply:
Huhhhhhhhhhh yeah… not sure how this wound up here! You are right about the intent, and that this wouldn’t work. I think I just need to drop this line and it will be correct. And unset `self.old_handlers` after so it can be garbage collected.